Excessive test time in some tests

#### Issue description

CI takes a loooong time for a variety of reasons but having all tests on critical path _and_ those tests taking a long time is one.  One can ask `catch2` to report times per test, and we can use this to make choices.

#### Your environment

 * version of mlpack: 4.5.2
 * operating system: Ubuntu 24.04
 * compiler: g++-14
 * version of dependencies (Boost/Armadillo): 'current'
 * any other environment information you think is relevant:

#### Steps to reproduce

1. Build with tests ie `-DBUILD_TESTS=ON`
2. Run `bin/mlpack_test -d yes | tee testTimings.txt` to create timing statistics
3. Sort them `grep ' s:' testTimings.txt | sort -n | tail -40`

#### Actual behavior

The fourty worst ones follow:

```sh
edd@rob:~/git/mlpack/build(master)$ grep ' s:' testTimings.txt | sort -n | tail -40
1.246 s: PerceptronSerializationTest - fmat
1.296 s: ClassifyTest_VERTEBRALCOL - fmat
1.385 s: PredictFloatTest
1.429 s: RegularizedSVDFunctionOptimizeHOGWILD
1.492 s: CartPoleWithNStepDQN
1.576 s: OneStepQLearningTest
1.622 s: PredictNormalizeDataTest
1.880 s: CartPoleWithDQNPrioritizedReplay
1.979 s: GradientBatchNormTest
2.179 s: LARSSparseModelDenseData - double
2.394 s: BiasSVDFunctionParallelOptimize
2.402 s: IncrementalTraining
2.666 s: OneStepSarsaTest
2.674 s: NCADifferentMaxIterationTest
2.833 s: LARSSparseModelDenseData - float
3.142 s: PendulumWithSAC
3.536 s: SVDBatchRegularizationTest - double
3.580 s: NCADifferentToleranceTest
3.610 s: SparseAutoencoderFunctionGradient
3.807 s: GradientGroupedConvolutionLayerTest
3.917 s: SVDBatchMomentumTest - float
4.047 s: NaiveGuaranteeTest
4.883 s: SVDBatchRegularizationTest - float
4.907 s: SparseCodingTrainReturnObjective - arma::fmat
4.938 s: SVDBatchMomentumTest - double
4.965 s: CFMinResidueTest
5.557 s: SparseCodingTrainReturnObjective - arma::mat
6.386 s: RBFNetworkTest
9.579 s: RAModelTest
17.137 s: PendulumWithGaussianDDPG
21.864 s: PendulumWithDDPG
22.812 s: OptimizerTest
24.396 s: NCADifferentNumBasisTest
45.529 s: PendulumWithTD3
45.941 s: ForwardBackwardTest
52.655 s: ExhaustiveSparseKDTreeTest
68.410 s: LargeRhoValueRnnTest
128.046 s: VanillaNetworkBatchSizeTest
454.358 s: VanillaNetworkTest
906.626 s: CheckCopyVanillaNetworkTest
edd@rob:~/git/mlpack/build(master)$ 
```

#### Hoped-for behavior

`VanillaNetworkTest` alone accounts for (907 + 454 + 128)/60 = 24.8 minutes which is the bulk of the test time.  I plan to take it behind the barn and shoot it into pieces.

Wait, what?  Your 24.10 test shows 7 minutes for the one test. That is the bad case.

Or you saying the different between 0.3.28+ds_2 (bad, in 24.10) and 0.3.29+ds+4 (better per you in 25.04) is the difference?  I see nuttin' in the Debian changelog for these two:

```
openblas (0.3.28+ds-4) unstable; urgency=medium

  * pkg-config-libsuffix.patch: new patch, fixes pkg-config for 64-bit flavours
    (Closes: #1087874)

 -- SÃ©bastien Villemot <sebastien@debian.org>  Wed, 20 Nov 2024 11:52:32 +0100

openblas (0.3.28+ds-3) unstable; urgency=medium

  * Team upload.

  [ Dandan Zhang <zhangdandan@loongson.cn> ]
  * Add upstream merged patch to fix undefined reference to
    blas_set_parameter (Closes: #1086535)

 -- Gianfranco Costamagna <locutusofborg@debian.org>  Fri, 08 Nov 2024 14:59:11 +0100

```

(from https://changelogs.ubuntu.com/changelogs/pool/universe/o/openblas/openblas_0.3.28+ds-4/changelog_

--- a/.ci/linux-steps.yaml
+++ b/.ci/linux-steps.yaml
@@ -24,7 +24,7 @@ steps:
     git clone --depth 1 https://github.com/mlpack/jenkins-conf.git conf
 
     sudo apt-get install --yes --allow-unauthenticated \
-    	libopenblas-dev libstb-dev libcereal-dev xz-utils ccache
+        libopenblas-dev libstb-dev libcereal-dev xz-utils ccache
 
     if [ "$BINDING" = "python" ]; then
       python -m pip install --upgrade pip
@@ -68,7 +68,9 @@ steps:
   displayName: 'Build'
 
 # Run tests via ctest.
-- script: cd build && CTEST_OUTPUT_ON_FAILURE=1 ctest -T Test .
+# The OMP_NUM_THREADS is to prevent thrashing on older versions of OpenBLAS
+# (0.3.26 and older) where OpenMP and pthreads aren't playing together well.
+- script: cd build && OMP_NUM_THREADS=1 CTEST_OUTPUT_ON_FAILURE=1 ctest -T Test .
   displayName: 'Run tests via ctest'
 
 # Ccache stats (two verbosity levels supported)

--- a/.ci/macos-steps.yaml
+++ b/.ci/macos-steps.yaml
@@ -41,7 +41,7 @@ steps:
       ccache --set-config "hash_dir=false" 
       ccache --show-config
       ccache --zero-stats
-    fi  
+    fi
 
   displayName: 'Install Build Dependencies'
 
@@ -60,7 +60,9 @@ steps:
   displayName: 'Build'
 
 # Run tests via ctest.
-- script: cd build && CTEST_OUTPUT_ON_FAILURE=1 ctest -T Test .
+# The OMP_NUM_THREADS is to prevent thrashing on older versions of OpenBLAS
+# (0.3.26 and older) where OpenMP and pthreads aren't playing together well.
+- script: cd build && OMP_NUM_THREADS=1 CTEST_OUTPUT_ON_FAILURE=1 ctest -T Test .
   displayName: 'Run tests via ctest'
 
 # Ccache stats (two verbosity levels supported

--- a/.ci/macos-steps.yaml
+++ b/.ci/macos-steps.yaml
@@ -60,9 +60,7 @@ steps:
   displayName: 'Build'
 
 # Run tests via ctest.
-# The OMP_NUM_THREADS is to prevent thrashing on older versions of OpenBLAS
-# (0.3.26 and older) where OpenMP and pthreads aren't playing together well.
-- script: cd build && OMP_NUM_THREADS=1 CTEST_OUTPUT_ON_FAILURE=1 ctest -T Test .
+- script: cd build && CTEST_OUTPUT_ON_FAILURE=1 ctest -T Test .
   displayName: 'Run tests via ctest'
 
 # Ccache stats (two verbosity levels supported

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -54,7 +54,7 @@ jobs:
         run: |
           mkdir -p ~/.R
           cp -vax .github/etc/R_Makevars_${{ runner.os }} ~/.R/Makevars
-          
+
       - name: Configure Ccache for R
         run: |
           ccache --set-config "sloppiness=include_file_ctime"
@@ -113,7 +113,10 @@ jobs:
 
       - name: Run tests via ctest
         run: |
-          cd build && CTEST_OUTPUT_ON_FAILURE=1 ctest -T Test .
+          # The OMP_NUM_THREADS is to prevent thrashing on older versions of
+          # OpenBLAS (0.3.26 and older) where OpenMP and pthreads aren't playing
+          # together well.
+          cd build && OMP_NUM_THREADS=2 CTEST_OUTPUT_ON_FAILURE=1 ctest -T Test .
 
       - name: Upload R packages
         uses: actions/upload-artifact@v4.4.0

--- a/README.md
+++ b/README.md
@@ -143,8 +143,9 @@ neural network, a compilation error will occur.
 
 ***Warning:*** older versions of OpenBLAS (0.3.26 and older) compiled to use
 pthreads may use too many threads for computation, causing significant slowdown.
-See the [test build guide](doc/user/install.md#build-tests) for more details and
-simple workarounds.
+OpenBLAS versions compiled with OpenMP do not suffer from this issue.  See the
+[test build guide](doc/user/install.md#build-tests) for more details and simple
+workarounds.
 
 See also:
 

--- a/README.md
+++ b/README.md
@@ -141,9 +141,10 @@ add `#define MLPACK_ENABLE_ANN_SERIALIZATION` before including `<mlpack.hpp>`.
 If you don't define `MLPACK_ENABLE_ANN_SERIALIZATION` and your code serializes a
 neural network, a compilation error will occur.
 
-***Warning:*** older versions of OpenBLAS (0.3.26 and older) may use too many
-threads for computation, causing significant slowdown.  See the [test build
-guide](doc/user/install.md#build-tests) for more details and simple workarounds.
+***Warning:*** older versions of OpenBLAS (0.3.26 and older) compiled to use
+pthreads may use too many threads for computation, causing significant slowdown.
+See the [test build guide](doc/user/install.md#build-tests) for more details and
+simple workarounds.
 
 See also:
 

--- a/README.md
+++ b/README.md
@@ -141,6 +141,10 @@ add `#define MLPACK_ENABLE_ANN_SERIALIZATION` before including `<mlpack.hpp>`.
 If you don't define `MLPACK_ENABLE_ANN_SERIALIZATION` and your code serializes a
 neural network, a compilation error will occur.
 
+***Warning:*** older versions of OpenBLAS (0.3.26 and older) may use too many
+threads for computation, causing significant slowdown.  See the [test build
+guide](doc/user/install.md#build-tests) for more details and simple workarounds.
+
 See also:
 
  * the [test program compilation section](doc/user/install.md#compiling-a-test-program)

--- a/doc/embedded/supported_boards.md
+++ b/doc/embedded/supported_boards.md
@@ -64,7 +64,7 @@ cmake \
 |-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
 | CORTEXA76 | [Bootlin toolchain link](https://toolchains.bootlin.com/releases_aarch64.html)      | [Sysroot and toolchain prefix](#cortexa76) | [Cortex A76 on Wikipedia](https://en.wikipedia.org/wiki/ARM_Cortex-A76)  |
 |-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
-| C906      | [Bootlin toolchain link](https://toolchains.bootlin.com/releases_riscv64-lp64d.html)| [Sysroot and toolchain prefix](#c906) | [C906 on riscv](https://www.riscfive.com/2023/03/09/t-head-xuantie-c906-risc-v/) |
+| C906      | [Bootlin toolchain link](https://toolchains.bootlin.com/releases_riscv64-lp64d.html)| [Sysroot and toolchain prefix](#c906) | [C906 on riscv](https://www.riscvschool.com/2023/03/09/t-head-xuantie-c906-risc-v/) |
 |---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
 | x280      | [Bootlin toolchain link](https://toolchains.bootlin.com/releases_riscv64-lp64d.html)| [Sysroot and toolchain prefix](#x280) | [x280 on SiFive](https://www.sifive.cn/api/document-file?uid=x280-datasheet) |
 |---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

--- a/doc/quickstart/cpp.md
+++ b/doc/quickstart/cpp.md
@@ -43,13 +43,13 @@ missing header files (e.g., `mlpack.hpp` missing in versions prior to 4.0).  To
 ensure compatibility with the latest mlpack features and examples, we recommend
 building mlpack from source, as explained in the [main README](../../README.md).
 
-***Warning:*** on Linux systems, older versions of OpenBLAS (0.3.26 and older)
-can over-use the number of cores on your system, causing slow execution of
-mlpack programs, especially mlpack's test suite.  To prevent this, set
-`OMP_NUM_THREADS` as detailed [in the test build
-guide](../user/install.md#build-tests), or install the `libopenblas-openmp`
-package on Ubuntu or Debian, or `openblas-openmp` package on Fedora.  Ubuntu
-24.04, Debian bookworm, and Fedora 41 and older are all affected by this issue.
+***Warning:*** on Ubuntu and Debian systems, older versions of OpenBLAS (0.3.26
+and older) can over-use the number of cores on your system, causing slow
+execution of mlpack programs, especially mlpack's test suite.  To prevent this,
+set `OMP_NUM_THREADS` as detailed [in the test build
+guide](../user/install.md#build-tests), or install the `libopenblas-openmp-dev`
+package on Ubuntu or Debian and remove `libopenblas-pthread-dev`.  Ubuntu 24.04,
+Debian bookworm, and older are all affected by this issue.
 
 ## Installing mlpack from vcpkg
 

--- a/doc/quickstart/cpp.md
+++ b/doc/quickstart/cpp.md
@@ -37,12 +37,19 @@ docker run -it mlpack/mlpack /bin/bash
 If you prefer to build mlpack from scratch, see the
 [main README](../../README.md).
 
-**Note for Ubuntu LTS Users**: The libmlpack-dev version in the Ubuntu LTS 
+***Note for Ubuntu LTS Users***: The libmlpack-dev version in the Ubuntu LTS
 repositories may not always be the latest. This can lead to issues, such as
-missing header files (e.g., `mlpack.hpp` missing in versions prior to 4.0).
-To ensure compatibility with the latest mlpack features and examples,
-we recommend building mlpack from source, as explained in the
-[main README](../../README.md).
+missing header files (e.g., `mlpack.hpp` missing in versions prior to 4.0).  To
+ensure compatibility with the latest mlpack features and examples, we recommend
+building mlpack from source, as explained in the [main README](../../README.md).
+
+***Warning:*** on Linux systems, older versions of OpenBLAS (0.3.26 and older)
+can over-use the number of cores on your system, causing slow execution of
+mlpack programs, especially mlpack's test suite.  To prevent this, set
+`OMP_NUM_THREADS` as detailed [in the test build
+guide](../user/install.md#build-tests), or install the `libopenblas-openmp`
+package on Ubuntu or Debian, or `openblas-openmp` package on Fedora.  Ubuntu
+24.04, Debian bookworm, and Fedora 41 and older are all affected by this issue.
 
 ## Installing mlpack from vcpkg
 

--- a/doc/user/install.md
+++ b/doc/user/install.md
@@ -60,6 +60,14 @@ If you plan to write mlpack programs, make sure you have a C++ compiler that
 supports C++17 available (this may not be automatically installed by the package
 manager).
 
+***Warning:*** on Linux systems, older versions of OpenBLAS (0.3.26 and older)
+can over-use the number of cores on your system, causing slow execution of
+mlpack programs, especially mlpack's test suite.  To prevent this, set
+`OMP_NUM_THREADS` as detailed [below](#build-tests), or install the
+`libopenblas-openmp` package on Ubuntu or Debian, or `openblas-openmp` package
+on Fedora.  Ubuntu 24.04, Debian bookworm, and Fedora 41 and older are all
+affected by this issue.
+
 ## Install from source
 
 If you only intend to use mlpack in a C++ program, it is not necessary to
@@ -232,6 +240,24 @@ The `mlpack_test` program uses the [Catch2](https://github.com/catchorg/Catch2)
 library for unit testing; this supports many options---you can see them with
 `mlpack_test -h`.
 
+***Warning:*** on Linux systems, older versions of OpenBLAS (0.3.26 and older)
+can over-use the number of cores on your system, causing slow execution of
+mlpack programs, especially mlpack's test suite.  To prevent this, set
+`OMP_NUM_THREADS` to half the number of cores on your system; so, for a system
+with 8 cores, run
+
+```sh
+OMP_NUM_THREADS=4 bin/mlpack_test
+```
+
+This workaround should be applied to any mlpack program where:
+
+ * Compilation with OpenMP is enabled (e.g. `-fopenmp` is used),
+ * Armadillo is using OpenBLAS,
+ * OpenBLAS is version 0.3.26 or older, and
+ * OpenBLAS is compiled against pthreads (the default on Ubuntu, Debian, and
+   Fedora).
+
 ## Compiling a test program
 
 Once mlpack is installed and available on the system, it is easy to compile a
@@ -274,6 +300,9 @@ prefer different compilation options.
 ```sh
 g++ -O3 -std=c++17 -o my_program my_program.cpp -lopenblas -fopenmp
 ```
+
+ - See also the warning on OpenBLAS versions and too many threads in [the
+   previous section](#build-tests).
 
 See [the Armadillo documentation](https://arma.sourceforge.net/faq.html#linking)
 for more information on linking Armadillo programs.

--- a/doc/user/install.md
+++ b/doc/user/install.md
@@ -60,13 +60,13 @@ If you plan to write mlpack programs, make sure you have a C++ compiler that
 supports C++17 available (this may not be automatically installed by the package
 manager).
 
-***Warning:*** on Linux systems, older versions of OpenBLAS (0.3.26 and older)
-can over-use the number of cores on your system, causing slow execution of
-mlpack programs, especially mlpack's test suite.  To prevent this, set
-`OMP_NUM_THREADS` as detailed [below](#build-tests), or install the
-`libopenblas-openmp` package on Ubuntu or Debian, or `openblas-openmp` package
-on Fedora.  Ubuntu 24.04, Debian bookworm, and Fedora 41 and older are all
-affected by this issue.
+***Warning:*** on Ubuntu and Debian systems, older versions of OpenBLAS (0.3.26
+and older) can over-use the number of cores on your system, causing slow
+execution of mlpack programs, especially mlpack's test suite.  To prevent this,
+set `OMP_NUM_THREADS` as detailed [in the test build
+guide](../user/install.md#build-tests), or install the `libopenblas-openmp-dev`
+package on Ubuntu or Debian and remove `libopenblas-pthread-dev`.  Ubuntu 24.04,
+Debian bookworm, and older are all affected by this issue.
 
 ## Install from source
 
@@ -241,10 +241,11 @@ library for unit testing; this supports many options---you can see them with
 `mlpack_test -h`.
 
 ***Warning:*** on Linux systems, older versions of OpenBLAS (0.3.26 and older)
-can over-use the number of cores on your system, causing slow execution of
-mlpack programs, especially mlpack's test suite.  To prevent this, set
-`OMP_NUM_THREADS` to half the number of cores on your system; so, for a system
-with 8 cores, run
+compiled to use pthreads can over-use the number of cores on your system,
+causing slow execution of mlpack programs, especially mlpack's test suite.
+OpenBLAS versions compiled with OpenMP do not suffer from this issue.  To work
+around this problem, set `OMP_NUM_THREADS` to half the number of cores on your
+system; so, for a system with 8 cores, run
 
 ```sh
 OMP_NUM_THREADS=4 bin/mlpack_test

